text/0000-deprecate-component-lifecycle-hook-args.md

```diff
@@ -0,0 +1,139 @@
+
+The reason to officially deprecate lifecycle hook arguments is not only about messaging, but also because providing these arguments imposes an unnecessary performance penalty to every component in your application even if the arguments are not used.
+
+To provide the arguments to the lifecycle hooks, Ember.js has to eagerly "reify" and save-off any passed-in attributes to allow diffing and construct several wrapper objects. In the few occasions where this logic is actually necessary, developers should be able to use programmatic patterns familiar to them and manually track changes as needed, as exemplified in the Transition Path section below.
+
+# Transition Path
+
+The transition path followed will be the standard one, which encompasses using the deprecation API to deprecate the feature and the related deprecation guide. While the lifecycle hooks share a deprecation identifier, they will be addressed in turn.
+
+### `didInitAttrs`
+
+Since this lifecycle hook is [already deprecated](http://emberjs.com/deprecations/v2.x/#toc_ember-component-didinitattrs), we suggest taking this chance to address two deprecations at the same time. Imagine you have a component that stores a timestamp when it's initialized for later comparison.
+
+Before:
+
+``` javascript
+Ember.Component.extend({
+  didInitAttrs({ attrs }) {
+    this.set('initialTimestamp', attrs.timestamp);
+  }
+});
+```
+After:
+
+``` javascript
+Ember.Component.extend({
+  init() {
+    this._super(...arguments);
+
+    this.set('initialTimestamp', this.get('timestamp'));
+  }
+});
+```
+### `didReceiveAttrs`
+
+Let's say you want to animate a map widget from the old coordinates to the new coordinates.
+
+Before:
+
+``` javascript
+Ember.Component.extend({
+  didReceiveAttrs({ oldAttrs, newAttrs }) {
+    if (oldAttrs && oldAttrs.coordinates !== newAttrs.coordinates) {
+      this.map.move({ from: oldAttrs.coordinates, to: newAttrs.coordinates });
+    }
+  }
+});
+```
+After:
+
+``` javascript
+Ember.Component.extend({
+  didReceiveeAttrs() {
+    let oldCoordinates = this.get('_previousCoordinates');
+    let newCoordinates = this.get('coordinates');
+
+    if (oldCoordinates && oldCoordinates !== newCoordinates) {
+      this.map.move({ from: oldCoordinates, to: newCoordinates });
+    }
+
+    this.set('_previousCoordinates', newCoordinates);
+  }
+});
+```
+### `didUpdateAttrs`
+
+This hook is very similar to `didReceiveAttrs`, except it only runs on re-renders and not the initial render.
+
+Before:
+
+``` javascript
+Ember.Component.extend({
+  didUpdateAttrs({ oldAttrs, newAttrs }) {
+    if (oldAttrs && oldAttrs.coordinates !== newAttrs.coordinates) {
+      this.map.move({ from: oldAttrs.coordinates, to: newAttrs.coordinates });
+    }
+  }
+});
+```
+After:
+
+``` javascript
+Ember.Component.extend({
+  didUpdateAttrs() {
```

Comment:
This would also need an

```javascript
init() {
  this._super(...arguments);
  this.set('_previousCoordinates', this.get('coordinates'));
}
```

no?